### ISSUE TITLE
GenAPI cannot decode records when reading from metadata

### ISSUE THREAD

**Issue by ericstj** [dotnet/sdk#32196]
### Describe the bug
GenAPI does not emit records correctly.

### To Reproduce
```c#
namepsace ns;
public record class Foo();
```
Compile this and point GenAPI to it.

Expect: Same output

Actual: 
```c#
namespace ns
{
    public record Foo
    {
        public Foo() { }
        protected Foo(Foo original) { }
        protected virtual System.Type EqualityContract { get { throw null; } }
        public virtual Foo <Clone>$() { throw null; }
        public virtual System.Boolean? Equals(Foo? other) { throw null; }
        public override System.Boolean? Equals(System.Object? obj) { throw null; }
        public override System.Int32 GetHashCode() { throw null; }
        public static System.Boolean? operator ==(Foo? left, Foo? right) { throw null; }
        public static System.Boolean? operator !=(Foo? left, Foo? right) { throw null; }
        protected virtual System.Boolean PrintMembers(System.Text.StringBuilder builder) { throw null; }
        public override System.String ToString() { throw null; }
    }
}
```

Related: https://github.com/dotnet/sdk/issues/32165

AFAICT there isn't anything preserved in metadata to indicate that a type is a record.  Perhaps we could use a heuristic -- look for all the members https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records.

**Comment by ViktorHofer** [comment]
Related: https://github.com/dotnet/roslyn/issues/67335

**Comment by ViktorHofer** [comment]
This seems to be important, especially when 8.0 packages will be added to SBRP. Keeping in 8.0 for now.

**Comment by ericstj** [comment]
Seems like I was wrong in my statement above - genapi does recognize the type as a record type - when it happens to be from source, or a record class.  It misses record structs when read from metadata - I filed https://github.com/dotnet/roslyn/issues/69326.  

After realizing this I thought that this bug might just be cosmetic - that we'd emit more than necessary but it would still compile.  This is not the case because there are members included that cannot be represented in C# (eg: `<Clone>$()`).

Perhaps we can just filter out all compiler generated members on record types.  I gave this a try manually and found that it wasn't quite enough - since it looks like the compiler doesn't annotate the public constructor with CompilerGenerated and defining it results in the compiler omitting the Deconstruct method (which does have the annotation).  We could omit the generated constructor - we'd have to identify it as the one constructor that doesn't call any others (`this(..)`).
